### DISCUSSION TITLE
How to run a collapsed cell without expanding it?

### DISCUSSION THREAD

**Discussion by xujc** [jupyterlab/jupyterlab#15770] [Q&A]
For example, I have a collapsed cell with a function. I need to select (click) it If I want to run it, but then it will expand.   
Is it possible to run it without expanding it?

<img width="500" alt="Screenshot 2024-02-08 at 08 21 08" src="https://github.com/jupyterlab/jupyterlab/assets/7146271/986f821c-56b3-4edc-a547-d5e476e9b7b8">

**Comment by Labeeb2339** [comment]
You can run a selected cell without using the mouse on the collapsed cell: move the notebook selection to it with the keyboard (or select it before collapsing), then use `Ctrl+Enter` (`notebook:run-cell`). That command runs the selected cell and does not advance to another cell.

The collapsed/hidden state is separate from execution, so running the selected cell does not require expanding its source first. If clicking the cell expands it in your version, use the command palette or keyboard navigation to select it instead. The command is documented here: https://jupyterlab.readthedocs.io/en/stable/user/commands.html#notebook-run-cell

For a plugin or custom shortcut, invoke the same `notebook:run-cell` command against the notebook tracker rather than simulating a click on the cell.

**Comment by rootkiller6788** [comment]
Yes — running a collapsed cell does not expand it. What expands it is the click itself.

When you collapse a code cell, JupyterLab removes the editor from the page and shows a thin strip with the first line of source. That strip is a click-to-expand control (its tooltip is literally *"Click to expand"*), so any click on the collapsed cell toggles it open — by design, and unrelated to running.

To run it while it stays collapsed, make it the **active** cell and run from the keyboard:
- `Ctrl+Enter` (`notebook:run-cell`) — runs the active cell and keeps it selected; it stays collapsed.
- `Shift+Enter` (`notebook:run-cell-and-select-next`) — runs it and moves to the next cell; the run cell also stays collapsed.

Execution never uncollapses a code cell: the run path in `NotebookActions` only switches the notebook back to command mode and never touches the cell's collapsed flag. (Markdown cells are the one exception — running a markdown cell renders and expands it.)

Two practical notes:
- Select the cell **before** collapsing it, or leave it active after collapsing, so `Ctrl+Enter` is always ready.
- If the collapsed cell is **not** already the active one, there is no built-in way to reach it without expanding: `↑`/`↓` and `J`/`K` deliberately skip cells whose input is collapsed, and clicking *is* the expand gesture. If "point at an already-collapsed cell and run it" is the workflow you want, that would be a reasonable feature request.
